cdap-api/src/main/java/co/cask/cdap/api/mapreduce/MapReduceContext.java

```diff
@@ -102,21 +104,44 @@
    * Overrides the output configuration of this MapReduce job to write to the specified dataset by its name.
    *
    * @param datasetName the name of the output dataset
+   * Deprecated as of version 3.2.0. Use {@link #addOutput(String)}, instead.
    */
+  @Deprecated
   void setOutput(String datasetName);
 
   /**
    * Overrides the output configuration of this MapReduce job to write to the specified dataset instance.
-   * Currently, the dataset passed in must either be an {@link OutputFormatProvider}.
+   * Currently, the dataset passed in must be an {@link OutputFormatProvider}.
    * You may want to use this method instead of {@link #setOutput(String)} if your output dataset uses runtime
    * arguments set in your own program logic.
    *
    * @param datasetName the name of the output dataset
    * @param dataset the output dataset
+   * Deprecated as of version 3.2.0. Use {@link #addOutput(String, Map)}, instead.
    */
+  @Deprecated
   void setOutput(String datasetName, Dataset dataset);
 
   /**
+   * Overrides the output configuration of this MapReduce job to also allow writing to the specified dataset by
+   * its name.
+   *
+   * @param datasetName the name of the output dataset
+   */
+  void addOutput(String datasetName);
+
+  /**
+   * Overrides the output configuration of this MapReduce job to also allow writing to the specified dataset.
+   * Currently, the dataset specified in must be an {@link OutputFormatProvider}.
```

Comment:
In the current framework, runtime arguments of a dataset aren't serialized into the hadoop conf and used when instantiating the dataset during task runtime.
Instead, they're used in OutputFormatProviders, to help the OutputFormatProvider figure out which configuration to return (for instance, in PFS/TPFS/FS).